### ISSUE TITLE
Tidb事务问题

### ISSUE THREAD

**Issue by xiebin2015** [pingcap/tidb#20717]
你好 PingCap团队，我尝试使用了TiDb试图测试下分布式事务，发现不起作用，望指正。
我搭建了一个TiDb的集群，
![image](https://user-images.githubusercontent.com/14512197/97549555-7da72300-1a0b-11eb-80c7-b7ac66ce9548.png)
写了个测试代码如图
![image](https://user-images.githubusercontent.com/14512197/97549686-a3342c80-1a0b-11eb-9f5d-50dd458f6722.png)
代码里面在第一个insert之后抛出异常，原本我期望的是数据库中不会新增记录，但事实上每次都新增了一条记录，了解到TiDb支持分布式事务，支持强一致性，请问是不是我哪里搞错了，请指教，谢谢

**Comment by zyguan** [comment]
Hi, @xiebin2015 , can you translate your code into sql, and also provide the schema of table as well as init data?

**Comment by bb7133** [comment]
> 代码里面在第一个insert之后抛出异常，原本我期望的是数据库中不会新增记录，但事实上每次都新增了一条记录

Hi @xiebin2015 , thanks for reporting!

May I know the detailed information about the Exception? Please provide the message and stack of it if possible.

**Comment by xiebin2015** [comment]
> Hi, @xiebin2015 , can you translate your code into sql, and also provide the schema of table as well as init data?

您好 ，感谢对我的回复
涉及的表很简单如下
SET NAMES utf8mb4;
SET FOREIGN_KEY_CHECKS = 0;

-- ----------------------------
-- Table structure for t_test
-- ----------------------------
DROP TABLE IF EXISTS `t_test`;
CREATE TABLE `t_test`  (
  `id` bigint(19) NOT NULL,
  `name` varchar(255) CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci NULL DEFAULT NULL,
  PRIMARY KEY (`id`) USING BTREE
) ENGINE = InnoDB CHARACTER SET = utf8mb4 COLLATE = utf8mb4_general_ci ROW_FORMAT = Compact;

SET FOREIGN_KEY_CHECKS = 1;

**Comment by xiebin2015** [comment]
> > 代码里面在第一个insert之后抛出异常，原本我期望的是数据库中不会新增记录，但事实上每次都新增了一条记录
> 
> Hi @xiebin2015 , thanks for reporting!
> 
> May I know the detailed information about the Exception? Please provide the message and stack of it if possible.
Hi  bb7133
您好，测试的代码很简单，就是在同一个方法里面试图往同一个表中插入两条记录，调用 第一个insert之后，故意抛出了一个异常来模拟事务失败的情况

    /**
     * @return
     * @throws PendingException
     */
    public ResponseDto<Test> save() throws PendingException {
        com.gf.tx.bo.test.Test test = new com.gf.tx.bo.test.Test();
        test.setId(IdGeneratorSingleton.getInstance().generateKey().longValue());
        test.setName("分布式事务测试-1");
        testMapper.insert(test);
        int i=1/0;
        com.gf.tx.bo.test.Test test1 = new com.gf.tx.bo.test.Test();
        test1.setId(IdGeneratorSingleton.getInstance().generateKey().longValue());
        test1.setName("分布式事务测试-111111");
        testMapper.insert(test1);
        return new ResponseDto<>();
    }
对于这段代码应该是涉及到了两个事务，所以当第一个insert调用之后就已经提交事务，然后再抛出异常，所以数据库里面是会多一条记录。
我说明一下我的理解：TiDB是一个分布式数据库，数据是分布在不同的数据节点上的，它解决的应该是在同一个事务中对不同数据分片的操作的强一致性，而不是解决对业务拆分之后的分布式事务的强一致性的。
举个例子：
有两个独立的微服务：A和B，现在有个业务W需要A通过RPC调用B的方式实现，这样W这个业务就形成了一个分布式事务（A有个事务、B也有事务，A\B都成功W才算成功），对于这个场景，TiDB是不解决的。
如果我的理解有偏差请指教，谢谢。

**Comment by zyguan** [comment]
> 代码里面在第一个insert之后抛出异常，原本我期望的是数据库中不会新增记录，但事实上每次都新增了一条记录，我想是因为这里实际上是有两个事务，所以第一个事务提交成功，导致数据库中新增了一条记录

Yes, I think you are right. It depends on whether `testMapper.insert(test);` gets committed or not by your client.

> 了解到TiDb支持分布式事务，支持强一致性，我理解的是它支持的是数据库层面的分布式事务，如果服务经过拆分，每个服务连接同一个TiDB，A服务的某个事务方法m1里面调用了B服务的方法m2，m1和m2各自有事务处理，这时候还能保障m1，和m2事务的一致性吗?

No, a transactions is associated with a connection to a tidb-server, if you have two services, they typically have their own connections (or connection pool), TiDB doesn't guarantee such a consistency, even they are connected to same tidb-server.

**Comment by xiebin2015** [comment]
> > 了解到TiDb支持分布式事务，支持强一致性，我理解的是它支持的是数据库层面的分布式事务，如果服务经过拆分，每个服务连接同一个TiDB，A服务的某个事务方法m1里面调用了B服务的方法m2，m1和m2各自有事务处理，这时候还能保障m1，和m2事务的一致性吗?
> 
> No, a transactions is associated with a connection to a tidb-server, if you have two services, they typically have their own connections (or connection pool), TiDB doesn't guarantee such a consistency, even they are connected to same tidb-server.

Yes, I see. Thank you.
